### PR TITLE
Simplify ConcreteIndices and its usage in TransportBulkAction

### PR DESCRIPTION
**Pull request by martijnvg**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -591,7 +591,7 @@ public void onResponse(BulkShardResponse bulkShardResponse) {
                     public void onFailure(Exception e) {
                         // create failures for all relevant requests
                         for (BulkItemRequest request : requests) {
-                            final String indexName = concreteIndices.getConcreteIndex(request.index()).getName();
+                            final String indexName = request.index();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:594]: No need to lookup `Index` from `ConcreteIndices` if we just use it as name in an exception message.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 2]
```diff
@@ -678,29 +678,18 @@ private boolean addFailureIfRequiresAliasAndAliasIsMissing(DocWriteRequest<?> re
...
-            final ConcreteIndices concreteIndices,
-            final Metadata metadata
-        ) {
+        private boolean addFailureIfIndexIsUnavailable(DocWriteRequest<?> request, int idx, final ConcreteIndices concreteIndices) {
             IndexNotFoundException cannotCreate = indicesThatCannotBeCreated.get(request.index());
             if (cannotCreate != null) {
                 addFailure(request, idx, cannotCreate);
                 return true;
             }
-            Index concreteIndex = concreteIndices.getConcreteIndex(request.index());
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:692]: The `ConcreteIndices#resolveIfAbsent(...)` already checks whether an `Index` instance has been cached, so
no need to this here as well.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 3]
```diff
@@ -738,10 +727,6 @@ void executeBulk(
             this.indexNameExpressionResolver = indexNameExpressionResolver;
         }
 
-        Index getConcreteIndex(String indexOrAlias) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:741]: After removing the above two usages, this method can be removed as it is no longer used.

**Review by csoulios — APPROVED** [review]
LGTM!

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:594]: I suspect that this one line could give some pretty big memory savings on the coordinating node sometimes. Prior to this change we'd retain a reference to the `ConcreteIndices` (and therefore the `ClusterState`) while the request was ongoing. Now we don't. Great stuff. 🎉

**Pull request merged**